### PR TITLE
[FLINK-28286][state] move 'enablechangelog' constant to flink-core module

### PR DESCRIPTION
**Pull request by fredia**
## What is the purpose of the change

Some methods in the `flink-runtime` module want to use `StreamConfig.ENABLE_CHANGE_LOG_STATE_BACKEND `constant(in `flink-streaming-java` module), but `flink-runtime` should not depend on `flink-streaming-java`.  This pr moves the changelog flag to `flink-core`.

See this [discussion](https://github.com/apache/flink/pull/19907#discussion_r902485780) for more details.

## Brief change log

- move `ENABLE_CHANGE_LOG_STATE_BACKEND` from `StreamConfig` to `org.apache.flink.api.common.ExecutionConfig`

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes,`org.apache.flink.api.common.ExecutionConfig` )
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java [hunk 1]
```diff
@@ -138,6 +139,8 @@ public class ExecutionConfig implements Serializable, Archiveable<ArchivedExecut
 
     private boolean isLatencyTrackingConfigured = false;
 
+    private TernaryBoolean changelogEnabled = TernaryBoolean.UNDEFINED;
+
```
Comment (rkhachatryan) [inline flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:143]: Have you considered moving the configuration to `JobInformation.jobConfiguration` instead of `ExecutionConfig`?

I think it would be easier for
- FLINK-26372 because `StateChangelogStorageFactory` requires multiple parameters
- and probably for FLINK-27692 if there will be more conifg parameters

WDYT?

cc: @zoltar9264

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java [hunk 2]
```diff
@@ -199,6 +199,7 @@ public StateBackend getStateBackend() {
 
     public void setChangelogStateBackendEnabled(TernaryBoolean changelogStateBackendEnabled) {
         this.changelogStateBackendEnabled = changelogStateBackendEnabled;
+        this.executionConfig.setChangelogStateBackendEnabled(changelogStateBackendEnabled);
```
Comment (rkhachatryan) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java:202]: I think `changelogStateBackendEnabled` field is unnecessary now and its read can be replaced with `executionConfig.isChangelogStateBackendEnabled()`.

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java (hunk 1 above)
Comment (fredia) [reply flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:143]: Yes, but there are two things to consider:
1. `JobInformation` is in `flink-runtime` module. Moving the configuration to `JobInformation.jobConfiguration` is not much different from the [current situation](https://github.com/apache/flink/pull/19907#discussion_r902485780) (`changelogEnabled ` is in `TaskInformation.taskConfiguration`). Do you mean moving the parsing to `JobInformation`? something like:

```java
public class JobInformation implements Serializable {
private static final String ENABLE_CHANGE_LOG_STATE_BACKEND = "enablechangelog";
  /** Configuration of the job. */
 private final Configuration jobConfiguration;
...
 public void setChangelogStateBackendEnabled(TernaryBoolean enabled) {
        try {
            InstantiationUtil.writeObjectToConfig(
                    enabled, this.jobConfiguration, ENABLE_CHANGE_LOG_STATE_BACKEND);
        } catch (Exception e) {
            throw new StreamTaskException(
                    "Could not serialize change log state backend enable flag.", e);
        }
    }
public TernaryBoolean isChangelogStateBackendEnabled(ClassLoader cl) {
        try {
            return InstantiationUtil.readObjectFromConfig(
                    this.jobConfiguration, ENABLE_CHANGE_LOG_STATE_BACKEND, cl);
        } catch (Exception e) {
            throw new StreamTaskException(
                    "Could not instantiate change log state backend enable flag.", e);
        }
    }
...
}
```
2. `BootstrapTransformation` in `flink-state-processing-api` also needs `enablechangelog` configuration, if move it  to `JobInformation.jobConfiguration`, do we need to keep a copy of this configuration in `flink-streaming-java`?

Comment (rkhachatryan) [reply flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:143]: I meant making `JobInformation.jobConfiguration` the **container** for changelog configuration; but having the constants (and the parsing code) in `flink-core`. That would be similar to `DistributedCache`.

`BootstrapTransformation` would be able to access the config too.

WDYT?

Comment (fredia) [reply flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:143]: The `periodicMaterializeIntervalMillis` and `materializationMaxAllowedFailures` associated with changelog are also in `ExecutionConfig`,  and I can't find a good class or a new class to put the constants (and the parsing code) in `flink-core`, do you have any suggestion?

**Review by rkhachatryan — APPROVED** [review]
Thanks for updating the PR, @fredia, LGTM.

Could you please rebase it and resolve conflicts?

cc:
@zoltar9264 (this might affect FLINK-26372 and FLINK-27155)
@zentol

**Comment by fredia** [comment]
Sure, Rebased. Thanks a lot for your review and patience!

File: flink-core/src/main/java/org/apache/flink/configuration/StateChangelogOptionsInternal.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+package org.apache.flink.configuration;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.util.FlinkRuntimeException;
+import org.apache.flink.util.InstantiationUtil;
+import org.apache.flink.util.TernaryBoolean;
+
+/** StateChangelog options that are used to pass job-level configuration from JM to TM. */
+@Internal
+public class StateChangelogOptionsInternal {
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/configuration/StateChangelogOptionsInternal.java:27]: This seems overly complicated. Why aren't you using `Configuration#getOptional`?

File: flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/BootstrapTransformation.java [hunk 4]
```diff
@@ -232,8 +231,6 @@ StreamConfig getConfig(
         config.setOperatorName(operatorID.toHexString());
         config.setOperatorID(operatorID);
         config.setStateBackend(stateBackend);
-        // This means leaving this stateBackend unwrapped.
-        config.setChangelogStateBackendEnabled(TernaryBoolean.FALSE);
```
Comment (zentol) [inline flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/BootstrapTransformation.java:236]: why is this removed?

Comment (zentol) [reply flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/BootstrapTransformation.java:236]: Aha, it's now directly encoded into the job configuration by the SavepointEnvironment.
